index.bs

```diff
@@ -16,13 +16,439 @@ WPT Display: closed
+-------------------------------------------------------------------------------------
+
+The <dfn http-header export>`Activate-Storage-Access`</dfn> HTTP response header
+allows a server to opt in to accessing its cookies in a cross-site request
+context. It is a [=Structured Field=] [=structured field/item=] whose value MUST be a [=structured
+field/token=]. [[!I-D.structured-field-values-for-http]] Its ABNF is:
+
+```
+Activate-Storage-Access = sf-item
+```
+
+Valid `Activate-Storage-Access` values include "`load`" and "`retry`". In order
+to support forward-compatibility with as-yet-unknown semantics, servers SHOULD
+ignore this header if it contains an invalid value.
+
+The following parameter is defined:
+
+* A parameter whose key is "`allowed-origin`", and whose value is a [=structured field/string=]. See below for processing requirements.
+
+<pre class="example" id="activate-storage-access-usage">
+
+// The server's response requests that the user agent activate storage access
+// before continuing with the load of the resource.
+Activate-Storage-Access: load
+
+// The server's response requests that the user agent activate storage access,
+// then retry the request. The "allowed-origin" parameter allowlists the
+// request's origin.
+Activate-Storage-Access: retry; allowed-origin="https://foo.bar"
+
+// Same as above, but using a wildcard instead of explicitly naming the request's Origin.
+Activate-Storage-Access: retry; allowed-origin=*
+</pre>
+
+<div algorithm>
+    To <dfn export>perform a storage access retry check</dfn> for a [=request=] |request| and [=response=] |response|, run the following steps:
+
+    <ol class="algorithm">
+        1. If |request|'s [=request/credentials mode=] [=string/is=] not "`include`", return failure.
+        1. If |request|'s [=request/eligible for storage-access=] is true, return failure.
+        1. Let |storageAccessStatus| be |request|'s [=request/storage access status=].
+        1. If |storageAccessStatus| [=string/is=] not "<code>[=storage access status/inactive=]</code>", return failure.
+        1. Let |parsedHeader| be the result of [=header list/get a structured field value|getting a structured field value=] given "`Activate-Storage-Access`" and "`item`" from |response|'s [=response/header list=].
+        1. If |parsedHeader| is null, return failure.
+        1. Let (|value|, |params|) be |parsedHeader|.
+        1. If |value| is not a Token, return failure.
+        1. If |value| [=string/is=] not "`retry`", return failure.
+        1. If |params|["allowed-origin"] does not exist, return failure.
+        1. Let |allowedOrigin| be |params|["allowed-origin"].
+        1. If |allowedOrigin| is a Token whose value is "`*`", return success.
+        1. If |allowedOrigin| is not a String, return failure.
+        1. Let |expectedOrigin| be the result of [=header list/get|getting=] "`Origin`" from |request|'s [=request/header list=].
```

Comment:
While it might work, I don't think string-comparing the origin header and the field value is good practice. Can you get the request's origin and `[=same origin=]` compare it to an origin parsed from the ASA header?

Reply:
> The problem with that is the request's origin is sometimes serialized as "null", even if it's a non-null (non-opaque) origin.

Are you referring to the redirect-tained condition, in https://fetch.spec.whatwg.org/#serializing-a-request-origin? In that case, I'm wondering if bypassing that condition in this spec is safe? Or do we handle the redirects correctly and safely, as Fetch would normally do?